gloo/transport/uv/pair.cc

```diff
@@ -0,0 +1,530 @@
+    readNextOp();
+  }
+
+  // Wake up connect function.
+  cv_.notify_one();
+}
+
+// Make next read operation read an operation's preamble struct.
+//
+// Threading: called from event loop thread.
+// Locking requirements: caller must hold instance mutex.
+//
+void Pair::readNextOp() {
+  // Reset pending operation.
+  readOp_ = Op();
+  // Next read must fill the preamble struct.
+  handle_->read((char*)&readOp_.preamble, sizeof(readOp_.preamble));
+}
+
+// Called if the handle is closed by the peer.
+//
+// Threading: called from event loop thread.
+// Locking requirements: none.
+//
+void Pair::onEnd(const libuv::EndEvent& event, const libuv::TCP&) {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    errno_ = UV_EOF;
+    handle_->close();
+    handle_.reset();
+  }
+  cv_.notify_all();
+}
+
+// Called if the handle saw an error.
+//
+// Threading: called from event loop thread.
+// Locking requirements: none.
+//
+void Pair::onError(const libuv::ErrorEvent& event, const libuv::TCP&) {
+  {
+    std::lock_guard<std::mutex> lock(mutex_);
+    errno_ = event.code();
+    handle_->close();
+    handle_.reset();
+  }
+  cv_.notify_all();
+}
+
+// Called on read completion.
+//
+// Threading: called from event loop thread.
+// Locking requirements: none.
+//
+void Pair::onRead(const libuv::ReadEvent& event, const libuv::TCP&) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  auto& op = readOp_;
+
+  // If this is the first read for the current operation,
+  // assert that we read the entire preamble.
+  if (op.nread == 0) {
+    GLOO_ENFORCE(event.length == sizeof(op.preamble));
+  }
+
+  // Capture total number of bytes read for the current operation.
+  op.nread += event.length;
+
+  const auto opcode = op.getOpcode();
+  if (opcode == Op::SEND_UNBOUND_BUFFER) {
+    // Remote side is sending data; find pending recv operation
+    // and read into the associated unbound buffer.
+    if (!op.buf) {
+      auto it = localPendingRecv_.find(op.preamble.tag);
+      GLOO_ENFORCE(it != localPendingRecv_.end());
+
+      // Ensure queue of pending recv operations is not empty.
+      auto& queue = it->second;
+      GLOO_ENFORCE(!queue.empty());
+
+      // Move pending recv operation to stack.
+      auto pendingRecv = std::move(queue.front());
+      queue.pop_front();
+      if (queue.empty()) {
+        localPendingRecv_.erase(it);
+      }
+
+      // Lock pointer to unbound buffer.
+      op.buf = NonOwningPtr<UnboundBuffer>(pendingRecv.buf);
+      GLOO_ENFORCE(op.buf, "Cannot lock pointer to unbound buffer");
+      op.offset = pendingRecv.offset;
+      op.length = pendingRecv.length;
+
+      // Read into unbound buffer, if the read is non-empty.
+      if (op.length) {
+        handle_->read((char*)op.buf->ptr + op.offset, op.length);
+        return;
+      }
+    }
+
+    GLOO_ENFORCE(op.nread == op.preamble.nbytes);
+    onSendUnboundBuffer(op);
+  } else if (opcode == Op::NOTIFY_SEND_READY) {
```

Comment:
What is the protocol here? When do we need `NOTIFY_SEND_READY ` and `NOTIFY_RECV_READY ` ops? Is there always a handshake before sending real `SEND_UNBOUND_BUFFER ` ops? If so, why? If the receiver can accept `NOTIFY_SEND_READY `, can't it just directly accept `SEND_UNBOUND_BUFFER`?

Reply:
This is done because the remote recv may not yet have been posted, OR it may be a recv from any operation. In either case, the unbound buffer is only sent when the pair knows there is a remote recv operation ready (i.e. it has received a `NOTIFY_RECV_READY` for that tag).

The `NOTIFY_SEND_READY` op is used to match against a recv from any on the other side. For example, if you have a single receiver and many senders (as is the case for RPC), the flow is as follows. The receiver side won't post any messages, just wait for `NOTIFY_SEND_READY` to arrive. The first one to arrive (again for the right tag and from the set of allowed ranks) will receive back a `NOTIFY_RECV_READY` message to ACK that it can send, and it will send the buffer. If then a few more `NOTIFY_SEND_READY` messages come in, the receiver can pick any of them to be the sender when the next recv from any call is made.

This mechanism is identical between the uv and tcp backends.